[LV] Use vscale for tuning when updating profile information

david-arm opened the pull request:
In fixVectorizedLoop we call setProfileInfoAfterUnrolling to update the profile information after vectorising, however for scalable VFs we pessimistically assume vscale=1. We can improve upon this by using the value of vscale used for tuning, i.e. when targeting neoverse-v1 the expected value is 2.

lukel97 reviewed (approved):
LGTM

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/check-prof-info.ll:
```diff
@@ -0,0 +1,123 @@
+
+; Check correctness of profile info for vectorization without epilog.
+; Function Attrs: nofree norecurse nounwind uwtable
+define dso_local void @_Z3foov() local_unnamed_addr #0 {
```
Would be good to clean up the test a bit, dropping `dso_local`, `local_unnamed_addr` ,` #0`

david-arm replied:
I just copied this from the existing test, which means we should also clean up the existing test too then. Happy to do that in this patch.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/check-prof-info.ll:
```diff
@@ -0,0 +1,123 @@
+  br label %for.body
+
+for.body:                                         ; preds = %for.body, %entry
+  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %for.body ]
```
```suggestion
  %iv = phi i64 [ 0, %entry ], [ %iv.next, %for.body ]
```

for consistency with other, newer tests

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/check-prof-info.ll:
```diff
@@ -0,0 +1,123 @@
+
+for.body:                                         ; preds = %for.body, %entry
+  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %for.body ]
+  %arrayidx = getelementptr inbounds [1024 x i32], ptr @b, i64 0, i64 %indvars.iv
+  %0 = load i32, ptr %arrayidx, align 4
+  %1 = trunc i64 %indvars.iv to i32
+  %mul = mul nsw i32 %0, %1
+  %arrayidx2 = getelementptr inbounds [1024 x i32], ptr @a, i64 0, i64 %indvars.iv
+  %2 = load i32, ptr %arrayidx2, align 4
+  %add = add nsw i32 %2, %mul
+  store i32 %add, ptr %arrayidx2, align 4
```
We don't really care about the body here right? Might be worth making it a bit simpler, maybe just with a store of the IV or load/store?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2787,10 +2801,10 @@ void InnerLoopVectorizer::fixVectorizedLoop(VPTransformState &State) {
   Loop *VectorLoop = LI->getLoopFor(HeaderBB);
-  setProfileInfoAfterUnrolling(OrigLoop, VectorLoop, OrigLoop,
-                               VF.getKnownMinValue() * UF);
+  unsigned VFxUF = getEstimatedRuntimeVF(VF * UF, Cost->getVScaleForTuning());
```
```suggestion
  unsigned EstimatedVFxUF = getEstimatedRuntimeVF(VF * UF, Cost->getVScaleForTuning());
```

Might be worth updating the name as well?

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/check-prof-info.ll:
```diff
@@ -0,0 +1,123 @@
+@b = dso_local global [1024 x i32] zeroinitializer, align 16
+
+; Check correctness of profile info for vectorization without epilog.
+; Function Attrs: nofree norecurse nounwind uwtable
```
```suggestion
```

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/check-prof-info.ll:
```diff
@@ -0,0 +1,123 @@
+@a = dso_local global [1024 x i32] zeroinitializer, align 16
+@b = dso_local global [1024 x i32] zeroinitializer, align 16
+
+; Check correctness of profile info for vectorization without epilog.
```
Might be good to spell spell out that we expect the branch weigth computations to use vscale = 1 for neoverse-v1  and  vscale = 2 for neoverse-v2?

david-arm committed:
[LV] Use vscale for tuning when updating profile information

In fixVectorizedLoop we call setProfileInfoAfterUnrolling to
update the profile information after vectorising, however
for scalable VFs we pessimistically assume vscale=1. We can
improve upon this by using the value of vscale used for
tuning, i.e. when targeting neoverse-v1 the expected value
is 2.

fhahn reviewed (commented):
LGTM, thanks!
